src/Vulkan/VkBuffer.cpp

```diff
@@ -0,0 +1,184 @@
+﻿#include "VkBuffer.h"
+
+#include <Vex/Buffer.h>
+
+#include "VkCommandQueue.h"
+#include "VkErrorHandler.h"
+#include "VkGPUContext.h"
+#include "VkMemory.h"
+
+namespace vex::vk
+{
+::vk::BufferUsageFlags GetVkBufferUsageFromDesc(const BufferDescription& desc)
+{
+    using enum ::vk::BufferUsageFlagBits;
+    switch (desc.usage)
+    {
+    case BufferUsage::StagingBuffer:
+        return eTransferSrc;
+    case BufferUsage::GenericBuffer:
+        return eStorageBuffer;
+    case BufferUsage::IndexBuffer:
+        return eIndexBuffer;
+    case BufferUsage::VertexBuffer:
+        return eVertexBuffer;
+    default:
+        VEX_LOG(Fatal, "RHIBuffer::Usage not supported by VulkanRHI")
+        break;
+    }
+    std::unreachable();
+}
+
+::vk::MemoryPropertyFlags GetMemoryPropsFromDesc(const BufferDescription& desc)
+{
+    using enum ::vk::MemoryPropertyFlagBits;
+
+    bool GPUWrite = desc.memoryAcces & BufferMemoryAccess::GPUWrite;
+    bool GPURead = desc.memoryAcces & BufferMemoryAccess::GPURead;
+    bool CPUWrite = desc.memoryAcces & BufferMemoryAccess::CPUWrite;
+    bool CPURead = desc.memoryAcces & BufferMemoryAccess::CPURead;
+
+    if (!CPUWrite && !CPURead)
+    {
+        if (GPUWrite || GPURead)
+        {
+            return eDeviceLocal;
+        }
+    }
+    else
+    {
+        if (GPUWrite || GPURead)
+        {
+            return eHostCoherent | eHostVisible;
+        }
+    }
+
+    std::unreachable();
+}
+
+VkBuffer::VkBuffer(VkGPUContext& ctx, const BufferDescription& desc)
+    : RHIBuffer{ desc }
+    , ctx{ ctx }
+{
+    auto bufferUsage = GetVkBufferUsageFromDesc(desc);
+    auto memoryProps = GetMemoryPropsFromDesc(desc);
+
+    if (memoryProps & ::vk::MemoryPropertyFlagBits::eDeviceLocal)
+    {
+        // Needs to get its data from somewhere. Will therefore always need a transfer dest usage
+        bufferUsage |= ::vk::BufferUsageFlagBits::eTransferDst;
+    }
+
+    buffer = VEX_VK_CHECK <<=
+        ctx.device.createBufferUnique({ .size = desc.size,
+                                        .usage = bufferUsage,
+                                        .sharingMode = ::vk::SharingMode::eExclusive,
+                                        .queueFamilyIndexCount = 1,
+                                        .pQueueFamilyIndices = &ctx.graphicsPresentQueue.family });
+
+    const ::vk::MemoryRequirements reqs = ctx.device.getBufferMemoryRequirements(*buffer);
+
+    memory = VEX_VK_CHECK <<= ctx.device.allocateMemoryUnique(
+        { .allocationSize = reqs.size,
+          .memoryTypeIndex = GetBestMemoryType(ctx.physDevice, reqs.memoryTypeBits, memoryProps) });
+
+    VEX_VK_CHECK << ctx.device.bindBufferMemory(*buffer, *memory, 0);
```

Comment:
As mentioned previously this is a good way to do this for now, eventually we should migrate to having a custom allocator strategy that would allow for memory reuse between resources, and reduce the total amount of allocations.